File: hphp/hack/test/ocamlrep_custom/test_custom.ml
(*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)

open Core_kernel

type drop_test (* abstract type for the Rust DropTest type *)

type ref_cell (* abstract type for the Rust Rc<RefCell<bool>> type *)

external drop_test_new : unit -> drop_test = "test_custom_drop_test_new"

external drop_test_custom_ref_count : drop_test -> int
  = "test_custom_drop_test_custom_ref_count"

external drop_test_get_cell : drop_test -> ref_cell
  = "test_custom_drop_test_get_cell"

external drop_test_cell_is_dropped : ref_cell -> bool
  = "test_custom_drop_test_cell_is_dropped"

let test_is_dropped_on_gc () =
  let drop_test = ref (Some (drop_test_new ())) in
  let cell = drop_test_get_cell (Option.value_exn !drop_test) in

  assert (Int.equal (drop_test_custom_ref_count (Option.value_exn !drop_test)) 1);
  assert (not @@ drop_test_cell_is_dropped cell);

  Gc.full_major ();
  assert (Int.equal (drop_test_custom_ref_count (Option.value_exn !drop_test)) 1);
  assert (not @@ drop_test_cell_is_dropped cell);

  drop_test := None;
  Gc.full_major ();
  assert (drop_test_cell_is_dropped cell);

  ()

type boxed_int (* abstract tyep for Rust BoxedInt type *)

external boxed_int_register : unit -> unit = "test_custom_boxed_int_register"

external boxed_int_new : int -> boxed_int = "test_custom_boxed_int_new"

external boxed_int_equal : boxed_int -> boxed_int -> bool
  = "test_custom_boxed_int_equal"

let test_boxed_int_serialize () =
  boxed_int_register ();

  let x = boxed_int_new 1 in
  assert (not @@ boxed_int_equal x (boxed_int_new 2));
  assert (boxed_int_equal x (boxed_int_new 1));

  let x_serialized = Marshal.to_bytes x [] in
  let x_deserialized = Marshal.from_bytes x_serialized 0 in
  Gc.full_major ();

  assert (boxed_int_equal x x_deserialized);
  assert (not @@ boxed_int_equal x_deserialized (boxed_int_new 2));
  Gc.full_major ();

  ()

let test_cases = [test_is_dropped_on_gc; test_boxed_int_serialize]

let main () = List.iter test_cases ~f:(fun test -> test ())

let () = main ()

File: hphp/hack/test/ocamlrep_custom/test_custom.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.
use ocamlrep_custom::caml_serialize_default_impls;
use ocamlrep_custom::{CamlSerialize, Custom};
use ocamlrep_ocamlpool::ocaml_ffi;

use std::cell::RefCell;
use std::convert::TryInto;
use std::io::Write;
use std::rc::Rc;

struct DropTest(Rc<RefCell<bool>>);

impl CamlSerialize for DropTest {
    caml_serialize_default_impls!();
}

struct DropTestCell(Rc<RefCell<bool>>);

impl CamlSerialize for DropTestCell {
    caml_serialize_default_impls!();
}

impl DropTest {
    pub fn new() -> Self {
        Self(Rc::new(RefCell::new(false)))
    }

    pub fn cell(&self) -> Rc<RefCell<bool>> {
        self.0.clone()
    }
}

impl Drop for DropTest {
    fn drop(&mut self) {
        *self.0.borrow_mut() = true;
    }
}

ocaml_ffi! {
  fn test_custom_drop_test_new() -> Custom<DropTest> {
    Custom::from(DropTest::new())
  }

  fn test_custom_drop_test_custom_ref_count(x: Custom<DropTest>) -> usize {
    let w = Rc::downgrade(x.inner());
    drop(x);
    w.strong_count()
  }

  fn test_custom_drop_test_get_cell(x: Custom<DropTest>) -> Custom<DropTestCell> {
    Custom::from(DropTestCell(x.cell()))
  }

  fn test_custom_drop_test_cell_is_dropped(x: Custom<DropTestCell>) -> bool {
    *x.0.borrow()
  }
}

struct BoxedInt(isize);

impl CamlSerialize for BoxedInt {
    caml_serialize_default_impls!();

    fn serialize(&self) -> Vec<u8> {
        let mut buffer = Vec::new();
        buffer.write_all(&self.0.to_be_bytes()).unwrap();
        buffer
    }

    fn deserialize(buffer: &[u8]) -> Self {
        let i = isize::from_be_bytes(buffer[0..std::mem::size_of::<isize>()].try_into().unwrap());
        BoxedInt(i)
    }
}

ocaml_ffi! {
  fn test_custom_boxed_int_register() {
    // Safety: called from OCaml in a single-threaded context.
    unsafe {
      BoxedInt::register();
    }
  }

  fn test_custom_boxed_int_new(x: isize) -> Custom<BoxedInt> {
    Custom::from(BoxedInt(x))
  }

  fn test_custom_boxed_int_equal(x: Custom<BoxedInt>, y: Custom<BoxedInt>) -> bool {
    x.0 == y.0
  }
}
